karapace/serialization.py

```diff
@@ -165,59 +170,70 @@ def __init__(
-        if schema_type is SchemaType.AVRO:
-            if isinstance(schema_typed.schema, avro.schema.NamedSchema):
-                namespace = schema_typed.schema.namespace
-        if schema_type is SchemaType.JSONSCHEMA:
-            namespace = schema_typed.to_dict().get("namespace", "dummy")
-        #  Protobuf does not use namespaces in terms of AVRO
-        if schema_type is SchemaType.PROTOBUF:
-            namespace = ""
-        return f"{self.subject_name_strategy(topic_name, namespace)}-{subject_type}"
+    def get_subject_name(
+        self, topic_name: str, schema: Union[ParsedTypedSchema, None], subject_type: str, schema_type: SchemaType
+    ) -> Subject:
+        namespace = None
+
+        if schema is not None:
+            namespace = "dummy"
+            if schema_type is SchemaType.AVRO:
+                if isinstance(schema.schema, avro.schema.NamedSchema):
+                    namespace = schema.schema.namespace
+            if schema_type is SchemaType.JSONSCHEMA:
+                namespace = schema.to_dict().get("namespace", "dummy")
+            #  Protobuf does not use namespaces in terms of AVRO
+            if schema_type is SchemaType.PROTOBUF:
+                namespace = ""
+        return Subject(f"{self.subject_name_strategy(topic_name, namespace)}-{subject_type}")
 
     async def get_schema_for_subject(self, subject: str) -> TypedSchema:
         assert self.registry_client, "must not call this method after the object is closed."
-
         schema_id, schema = await self.registry_client.get_latest_schema(subject)
         async with self.state_lock:
             schema_ser = str(schema)
             self.schemas_to_ids[schema_ser] = schema_id
             self.ids_to_schemas[schema_id] = schema
         return schema
 
-    async def get_id_for_schema(self, schema: str, subject: str, schema_type: SchemaType) -> int:
+    async def has_schema_id_on_subject(self, subject: str, schema_id: SchemaId) -> bool:
         assert self.registry_client, "must not call this method after the object is closed."
-        try:
-            schema_typed = ParsedTypedSchema.parse(schema_type, schema)
-        except InvalidSchema as e:
-            raise InvalidPayload(f"Schema string {schema} is invalid") from e
+        registered = await self.registry_client.subject_has_schema_id_registered(subject, schema_id)
+        return registered
+
+    async def get_id_for_schema(self, schema_typed: ValidatedTypedSchema, subject: str) -> SchemaId:
+        assert self.registry_client, "must not call this method after the object is closed."
+
         schema_ser = str(schema_typed)
+
+        # if it's None shouldn't we firstly query to the original source? Maybe the schema has been added
+        # without the rest proxy in between and so the map isn't updated!
         if schema_ser in self.schemas_to_ids:
             return self.schemas_to_ids[schema_ser]
+
+        # should we wrap the SchemaRetrievalError erro type?
```

Comment:
suggestion: I think either address these comments, or make them todos, or remove them.

Reply:
I've already fixed this in my last commit, I resolved my doubt about this 😄 and I've removed a branch from the get_schema algorithm